### PR TITLE
aAmpl Implementation

### PR DESCRIPTION
**Pull request by aalavandhan**
A proposed integration for the AMPL token for the Aave protocol.

AMPL total supply and addresses balances change once every 24 hours(at 2am UTC) based on a price oracle input. The supply change factor is the same for all holders, this is achieved by updating an internal factor used to calculate balanceOf(address) named _gonsPerFragment where gons is internal supply/balance representation https://github.com/ampleforth/uFragments/blob/212d341f03cd16c4ec50685c788e1dd807eeeded/contracts/UFragments.sol#L169 

Because of that, lending and borrowing AMPL can behave differently from fixed supply tokens.

The desired behavior is for the borrowed token count (loan) to stay the same and not get affected by rebase. And the unborrowed amount to get affected by rebase while deposited in the AAVE protocol.

The aAMPL token is modified from generic AToken to achieve that behavior. The AMPL debt is denominated in fixed supply units and AMPL deposits are 'partially' elastic. such that the borrowed amount stays the same while the unborrowed amount gets scaled with rebase of AMPL token

### AAmplToken

The AMPL AToken functionally behaves similarly to every other aToken. It always maintains a 1:1 peg with the underlying AMPL. It should guarantee the following to always be true.

## Example Scenarios
https://colab.research.google.com/drive/1a4zd7UL-U5Xrme9b0X5B39nJzjrFwfG1?usp=sharing

```
1) At any time, user can deposit x AMPLs to mint x aAMPLs. 
   Total aAMPL supply increases by exactly x.

2) At any time, user can burn x aAMPLs for x AMPLs.
   Total aAMPL supply decreases by exactly x.

3) At any time, userA can transfer x aAMPLs to userB.
   userA's aAMPL balance reduces by X.
   userB's aAMPL balance increases by X.
   Total aAMPL supply exactly remains same.

4) When AMPL's supply rebases, only the 'unborrowed' aAMPL should rebase.
  => Say there are 1000 aAMPL, and 200 AMPL is lent out. AMPL expands by 10%.
     The new aAMPL supply should be 1080 aAMPL.
  => Say there are 1000 aAMPL, and 200 AMPL is lent out. AMPL contracts by 10%.
     The new aAMPL supply should be 920 aAMPL.

5) When AMPL's supply rebases, only the part of the balance of a user proportional to  the available liquidity ('unborrowed') should rebase.
  => Say a user has deposited 1000 AMPL and receives 1000 aAMPL, and 
     20% of the total underlying AMPL is lent out. AMPL expands by 10%.
     The new aAMPL user balance should be 1080 aAMPL.
  => Say a user has deposited 1000 AMPL and receives 1000 aAMPL, and 
     20% of the total underlying AMPL is lent out. AMPL contracts by 10%.
     The new aAMPL supply should be 920 aAMPL.
```

***

## Implementation 
We implement a custom `AToken`, `StableDebtToken` and `VariableDebtToken` for AMPL. The `StableDebtToken` and `VariableDebtToken` implementations make no functional changes to the generic implementation. When the debt tokens are mint and burnt, it performs some additional book-keeping to keep track of the `totalScaledAMPL` (fixed AMPL denomination, ie `Gons`) borrowed at any-time. The new `getAMPLBorrowData` method returns the `totalScaledAMPL` and `totalScaledSupply` which are used for `AAmplToken` math.

The AAmplToken inherits from `IncentivizedAAmplERC20` rather than the `IncentivizedERC20` which the generic token inherits from. The `IncentivizedAAmplERC20` implements custom math to handle elastic aToken balances. The AAmplToken additionally stores references to the STABLE_DEBT_TOKEN, and the VARIABLE_DEBT_TOKEN contracts to calculate the total Principal borrowed (`totalAMPLBorrowed`, `totalScaledAMPLBorrowed`), at any time.

### Generic aToken Math

Generic AToken has a private balance and public balance. The public balance returned by `balanceOf` is multiplied by an interest rate factor. 

`_balances[u]` and `_totalSupply` hold the balance and total supply in storage.

```
AToken(u) = ATokenInternal(u) . I
AToken(supply) = ATokenInternal(supply) . I

ATokenInternal(u) = _balances[u]
ATokenInternal(supply) = _totalSupply
``` 

For example, a user deposits 100 DAI to get 100 aDAI and accrues 5 DAI interest. 
```
ADaiInternal(user) = 100
ADai(user) = 100 * 1.05 = 105 
```

###  aAMPL math

The aAMPL works the same way but the `AAmplInternal` balance is multiplied by another factor, which accounts for the underlying AMPL rebasing and the amount of AMPL unborrowed. 

```
AAmpl(u) = AAmplInternal(u) . I
AAmpl(supply) = AAmplInternal(supply) . I

# AMPL_SCALAR is inferred from the AMPL ERC-20 contract
AAmplInternal(supply) = (totalScaledAMPLDeposited - totalScaledAMPLBorrowed)/AMPL_SCALAR + totalAMPLBorrowed

aAMPL_Scalar() = AAmplInternal(supply) / _totalSupply
AAmplInternal(u) = aAMPL_Scalar() * _balances[u]
```

### CODE REVIEW

File: contracts/protocol/tokenization/ampl/AAmplToken.sol [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+     userA's aAMPL balance reduces by X.
+     userB's aAMPL balance increases by X.
+     Total aAMPL supply exactly remains same.
+
+  4) When AMPL's supply rebases, only the 'unborrowed' aAMPL should rebase.
+     => Say there are 1000 aAMPL, and 200 AMPL is lent out. AMPL expands by 10%.
+        Thew new aAMPL supply should be 1080 aAMPL.
+     => Say there are 1000 aAMPL, and 200 AMPL is lent out. AMPL contracts by 10%.
+        Thew new aAMPL supply should be 920 aAMPL.
+*/
```
Comment (aalavandhan) [inline contracts/protocol/tokenization/ampl/AAmplToken.sol]:
```suggestion

    5) When AMPL's supply rebases, only the part of the balance of a user proportional to  the available liquidity ('unborrowed') should rebase.
      => Say a user has deposited 1000 AMPL and receives 1000 aAMPL, and 
         20% of the total underlying AMPL is lent out. AMPL expands by 10%.
         The new aAMPL user balance should be 1080 aAMPL.
      => Say a user has deposited 1000 AMPL and receives 1000 aAMPL, and 
         20% of the total underlying AMPL is lent out. AMPL contracts by 10%.
         The new aAMPL supply should be 920 aAMPL.
*/
```

File: contracts/protocol/tokenization/ampl/IncentivizedAAmplERC20.sol [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+  }
+
+
+  /**
+   * @return The total supply of the token
+   *
+   *  totalSupply() = unborrowed + borrowed
+   *                = (totalScaledAMPLDeposited - totalScaledAMPLBorrowed)/AMPL_SCALAR + totalAMPLBorrowed
+   **/
+  function totalSupply() public view virtual override returns (uint256) {
```
Comment (nithinkrishna) [inline contracts/protocol/tokenization/ampl/IncentivizedAAmplERC20.sol:67]: `getAMPLScalar()` is called multiple times

File: contracts/protocol/tokenization/ampl/IncentivizedAAmplERC20.sol [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    remainingAccountBalance = totalSupply() - balanceOf(account)
+
+    Thus:
+    => _mintAmount = (_totalSupply * newAccountBalance) - (_balances[account] * newTotalSupply) / remainingAccountBalance
+  */
+  function _mint(address account, uint256 amount) internal virtual {
+    require(account != address(0), 'AAmplERC20: mint to the zero address');
+
+    _beforeTokenTransfer(address(0), account, amount);
+
```
Comment (aalavandhan) [inline contracts/protocol/tokenization/ampl/IncentivizedAAmplERC20.sol]: `getAMPLScalar()` is called multiple times  in the mint, burn and transfer functions

These functions internally call `totalSupply` and `balanceOf` which calll`getAMPLScalar()`

File: contracts/protocol/tokenization/ampl/IncentivizedAAmplERC20.sol [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+    address sender,
+    address recipient,
+    uint256 amount
+  ) internal virtual {
+    require(sender != address(0), 'AAmplERC20: transfer from the zero address');
+    require(recipient != address(0), 'AAmplERC20: transfer to the zero address');
+
+    _beforeTokenTransfer(sender, recipient, amount);
+
+    uint256 currentTotalSupply = totalSupply();
```
Comment (nithinkrishna) [inline contracts/protocol/tokenization/ampl/IncentivizedAAmplERC20.sol:196]: getAMPLScalar() is called multiple times here too

File: contracts/protocol/tokenization/ampl/AAmplToken.sol [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+      .mul(userBalanceInternalBefore)
+      .sub(totalSupplyInternalBefore.mul(userBalanceScaledAfter))
+      .div(otherBalanceScaledBefore);
+
+    _burn(user, burnAmountInternal);
+  }
+
+  /**
+   * @dev balanceOfScaled = balanceInternal / totalSupplyInternal * totalSupplyScaled
+   *                       = balanceInternal . λ
```
Comment (aalavandhan) [inline contracts/protocol/tokenization/ampl/AAmplToken.sol]:
```suggestion
   *                       = λ . balanceInternal
```

**Pull request merged**